### PR TITLE
Use the same resource validation logic for both pipelineruns + taskruns

### LINKED ISSUE #213: PipelineRun Controller Validation - Output Binding Issue
## Expected Behavior
Currently there is a bug in the PipelineRun Controller Validation as noted in https://github.com/knative/build-pipeline/pull/192#discussion_r229856526.  This allows for Outputs to not have proper Resource bindings.  A test for correctly working correct-output-source-bindings (in addition to the bad-output-source-bindings test we currently have) should catch this issue as currently nothing is added to outMapping here:
https://github.com/knative/build-pipeline/blob/master/pkg/reconciler/v1alpha1/pipelinerun/validate.go#L99

## Actual Behavior
Outputs mappings should be checked in a PipelineRun.  Currently, as nothing is added to the outMapping, it is not possible to have a verified PipelineRun that has Resource mapping

## Steps to Reproduce the Problem

  1. Create a Resource
  2. Create a Task requiring an Output with a ResourceRef
  3. Create a TaskRun referencing the created Resource

### PR DESCRIPTION
**Pull request by bobcatfish**
Now when resolving a PipelineRun, we will resolve all of the resource references as well. To do this, we make it so that PipelineRuns and TaskRuns use the same types to bind resources.

Then we use exactly the same logic that was used to validate taskrun resources to validate pipelinerun resources.

The structure that contains the "resolved" pipelinerun can probably be iterated on (some of the tests only need a subset of it, which is a hint that we probably shouldnt pass the whole thing around everywhere) but this should help with #168 b/c to implement a DAG we should just need to put the resolved objects into a DAG instead of a list.

Bonuses:
* pipelineruns will now respect tasks that have defaults for their resources
* users can provide resource paths in pipelineruns if they want to (which before only worked for taskruns)

Fixes #213

There is some future work we can do to use these resolved objects in the rest of the reconcilers as well (e.g. when applying templating) but I think we can iterate on it!

### CODE REVIEW

**Comment by bobcatfish** [comment]
```
I1203 04:39:10.739] --- FAIL: TestPipelineRun (0.00s)
I1203 04:39:10.739]     --- FAIL: TestPipelineRun/service_account_propagation (600.23s)
I1203 04:39:10.740]     	pipelinerun_test.go:157: Error waiting for PipelineRun helloworld-pipelinerun1 to finish: timed out waiting for the condition
I1203 04:39:10.740]     --- FAIL: TestPipelineRun/fan-in_and_fan-out (600.23s)
I1203 04:39:10.741]     	pipelinerun_test.go:157: Error waiting for PipelineRun helloworld-pipelinerun1 to finish: timed out waiting for the condition
```

ruh roh

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go [hunk 1]
```diff
@@ -186,21 +186,40 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1alpha1.PipelineRun) er
...
+				Status:  corev1.ConditionFalse,
+				Reason:  ReasonFailedValidation,
+				Message: err.Error(),
+			})
+			return nil
+		}
+	}
+	err = resources.ResolveTaskRuns(c.taskRunLister.TaskRuns(pr.Namespace).Get, pipelineState)
+	if err != nil {
+		return fmt.Errorf("error getting TaskRunss for Pipeline %s: %s", p.Name, err)
```
Comment (pivotal-nader-ziada) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: typo "TaskRunss"

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go [hunk 2]
```diff
@@ -186,21 +186,40 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1alpha1.PipelineRun) er
...
+			pr.Status.SetCondition(&duckv1alpha1.Condition{
+				Type:    duckv1alpha1.ConditionSucceeded,
+				Status:  corev1.ConditionFalse,
+				Reason:  ReasonFailedValidation,
+				Message: err.Error(),
+			})
+			return nil
+		}
+	}
+	err = resources.ResolveTaskRuns(c.taskRunLister.TaskRuns(pr.Namespace).Get, pipelineState)
```
Comment (pivotal-nader-ziada) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: I feel this func is not really doing much, but just here for the test!

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go [hunk 3]
```diff
@@ -194,8 +205,8 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error
 		return nil
 	}
 
-	if err := ValidateTaskRunAndTask(tr.Spec.Inputs.Params, rtr); err != nil {
-		c.Logger.Error("Failed to validate taskrun %s with error %v", tr.Name, err)
+	if err := ValidateResolvedTaskResources(tr.Spec.Inputs.Params, rtr); err != nil {
```
Comment (pivotal-nader-ziada) [inline pkg/reconciler/v1alpha1/taskrun/taskrun.go]: can `validateResolvedTaskResources` be part of `ResolveTaskResources`? we don't want to get back invalid resources

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go (hunk 1 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: whoops thanks :sweat_smile:

(it's `PipelineParamses` all over again XD)

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go (hunk 2 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: I felt kind of weird about it too! This logic used to be inside of `ResolvePipelineRun` - I think it makes sense to break it out but I'm not sure happy with how it looks right now :S Do you have any other ideas?

I'm thinking we might want to create a type that hold the `PipelineState` and have functions like this operate on it maybe :J that's the only idea I have right now

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 3 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: I really liked keeping these separate tbh! The logic required to follow all the references and find the objects I think is pretty different from the logic required to make sure all the required resources + params are present.

And we can have two different errors, one where you referred to something that couldn't be retrieved, and one where you didn't provide something that was needed

**Comment by bobcatfish** [comment]
There was a bug that the end to end tests caught! :D

Ready for another look @pivotal-nader-ziada

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go (hunk 2 above)
Comment (pivotal-nader-ziada) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: Having some internal types for ResolvedTasks would be useful

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 3 above)
Comment (pivotal-nader-ziada) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: fair point about different errors, we can always optimize later if we find out we need to

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go (hunk 2 above)
Comment (pivotal-nader-ziada) [reply pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: we can iterate on that one, but lets merge it for now

**Pull request merged**